### PR TITLE
Implement register_stratification, underlying functionality, and add tests

### PR DESCRIPTION
**Pull request by mattkappel**
### Description
- *Category*: POC
- *JIRA issue*: [MIC-3133](https://jira.ihme.washington.edu/browse/MIC-3133)

#### Changes
- Implements ResultsInterface and ResultsManager register_stratification along with ResultsContext functionality
- Refactor ResultsInterface to be separate from ResultsManager, using `TYPE_CHECKING` to handle circular imports
- Refactor often-used mocks to its own file
- Adds tests for register_stratification with columns and pipelines as sources
- TODO added for error case (duplicate stratification) pending feedback

### Testing
Four new tests were added, all pass along with the pre-existing ones for the results system.

### CODE REVIEW

File: src/vivarium/framework/results/interface.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    stratifications required by the model using :func:`register_default_stratifications`,
+    :func:`register_stratification`, and :func:`register_binned_stratification`
+    as necessary. A “binned stratification” is a stratification special case for
+    the very common situation when a continuous value needs to be binned into
+    categorical bins. The `is_vectorized` argument should be True if the mapper
+    function expects a DataFrame, and False if it expects a row of the DataFrame
+    and should be used by calling df.apply.
+    """
+
+    def __init__(self, manager: "ResultsManager") -> None:
```
Comment (stevebachmeier) [inline src/vivarium/framework/results/interface.py:42]: I think with TYPE_CHECKING implemented you can lose the quotes? I might be wrong...still kinda confused between using quotes in type-checking and using the TYPE_CHECKING import.

**Review by albrja — APPROVED** [review]

**Comment by albrja** [comment]
Nice work on the tests.  Helped my understanding of what was going on here.

File: src/vivarium/framework/results/interface.py (hunk 1 above)
Comment (mattkappel) [reply src/vivarium/framework/results/interface.py:42]: Good question, I think you need quotes for the `TYPE_CHECKING=False` case. https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

**Comment by mattkappel** [comment]
I've added duplicate Stratification handling in the the context and a test. This PR is ready for full review. I would appreciate in particular review from @rmudambi and @ramittal. 

Please pay special attention to:

- Changes in the ResultsManager src/vivarium/framework/results/manager.py
- Changes in the ResultsContext src/vivarium/framework/results/context.py
- Additions to the test suite in tests/framework/results/test_context.py and tests/framework/results/test_manager.py

Please ignore:

- Refactored out ResultsInterface from the manager file: src/vivarium/framework/results/interface.py
-  Refactored out mocks from pre-existing tests: tests/framework/results/mocks.py and tests/framework/results/test_stratification.py

File: src/vivarium/framework/results/manager.py [hunk 2]
```diff
@@ -38,7 +40,7 @@ def metrics(self):
         return self._metrics.copy()
 
     # noinspection PyAttributeOutsideInit
-    def setup(self, builder):
+    def setup(self, builder: "Builder"):
```
Comment (albrja) [inline src/vivarium/framework/results/manager.py:43]: Do not need the "s

File: src/vivarium/framework/results/manager.py [hunk 3]
```diff
@@ -130,6 +159,8 @@ def register_observation(
         )
 
     def _add_resources(self, target: List[str], target_type: str):
+        if not len(target):
+            return  # do nothing on empty lists
```
Comment (albrja) [inline src/vivarium/framework/results/manager.py:163]: Curious on checks for empty lists/dataframes/etc and good practices for that.

File: src/vivarium/framework/results/manager.py (hunk 2 above)
Comment (mattkappel) [reply src/vivarium/framework/results/manager.py:43]: Yes, you do.

File: src/vivarium/framework/results/manager.py (hunk 3 above)
Comment (mattkappel) [reply src/vivarium/framework/results/manager.py:163]: Here, it's not strictly necessary, but it's a low-cost check as most `target` lists will be of length 1. Good practice is to consider the tradeoff of the performance hit of the check vs the benefit.

File: src/vivarium/framework/results/interface.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    backend by means of the builder object. It exposes methods
+    to register stratifications, set default stratifications, and register
+    results producers. There is a special case for stratifications generated
+    by binning continuous data into categories.
+
+    The expected use pattern would be for a single component to register all
+    stratifications required by the model using :func:`register_default_stratifications`,
+    :func:`register_stratification`, and :func:`register_binned_stratification`
+    as necessary. A “binned stratification” is a stratification special case for
+    the very common situation when a continuous value needs to be binned into
```
Comment (rmudambi) [inline src/vivarium/framework/results/interface.py:36]: Nit: change to "... very common situation when a SINGLE continuous value..."

File: src/vivarium/framework/results/interface.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    categorical bins. The `is_vectorized` argument should be True if the mapper
+    function expects a DataFrame, and False if it expects a row of the DataFrame
```
Comment (rmudambi) [inline src/vivarium/framework/results/interface.py:38]: Maybe be a bit more explicit and say "The `is_vectorized` argument should be True if the mapper function expects a DataFrame corresponding to the whole population, and False if it expects a row of the DataFrame corresponding to a single simulant."

File: src/vivarium/framework/results/interface.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    and should be used by calling df.apply.
+    """
+
+    def __init__(self, manager: "ResultsManager") -> None:
+        self._manager: "ResultsManager" = manager
```
Comment (rmudambi) [inline src/vivarium/framework/results/interface.py:43]: Above we do typing with comments e.g. `self._default_stratifications = []  # type: List[str]`
We should decide on a consistent pattern.

File: src/vivarium/framework/results/manager.py [hunk 7]
```diff
@@ -82,12 +85,38 @@ def register_stratification(
...
+
+        Returns
+        ------
+        None
+        """
         target_columns = list(requires_columns) + list(requires_values)
         self._results_context.add_stratification(
             name, target_columns, categories, mapper, is_vectorized
         )
+        self._add_resources(requires_columns, "column")
```
Comment (rmudambi) [inline src/vivarium/framework/results/manager.py:118]: Should we use Enums rather than strings here? I tend not to like naked strings floating around.

**Review by rmudambi — COMMENTED** [review]
Only looked at the src directory so far. Everything looks good with a few minor nits and some suggestions. Looking at the tests now

File: tests/framework/results/test_context.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            SOURCES,
+            CATEGORIES,
+            sorting_hat_serial,
+            False,
+        ),
+    ],
+)
+def test_add_stratification(name, sources, categories, mapper, is_vectorized):
+    ctx = ResultsContext()
+    ctx.add_stratification(name, sources, categories, mapper, is_vectorized)
```
Comment (rmudambi) [inline tests/framework/results/test_context.py:39]: Maybe `assert expected_object not in ctx._stratifications` before adding it?

File: src/vivarium/framework/results/context.py [hunk 9]
```diff
@@ -42,11 +45,33 @@ def add_stratification(
...
+        is_vectorized
+            `True` if the mapper function expects a `DataFrame`, and `False` if it
+            expects a row of the `DataFrame` and should be used by calling :func:`df.apply`.
+
+
+        Returns
+        ------
+        None
+        """
+        if len([s.name for s in self._stratifications if s.name == name]):
```
[the remark this answers is not in the document]
Comment (rmudambi) [reply src/vivarium/framework/results/context.py:71]: Changing equality has implications for your tests below, so it's probably a bad idea to do this.

File: tests/framework/results/test_context.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    expected_object = Stratification(name, sources, categories, mapper, is_vectorized)
+    assert expected_object in ctx._stratifications
+
+
+@pytest.mark.parametrize(
+    "name, sources, categories, mapper, is_vectorized, expected_exception",
+    [
+        (  # map to a category that isn't in CATEGORIES
```
Comment (rmudambi) [inline tests/framework/results/test_context.py:47]: I think each of these cases should be a separate test, since they have nothing in common other than the fact that they should fail

File: tests/framework/results/test_context.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            NAME,
+            SOURCES,
+            CATEGORIES,
+            sorting_hat_bad_mapping,
+            False,
```
Comment (rmudambi) [inline tests/framework/results/test_context.py:52]: I'm very surprised this throws an exception. How do we know when adding a stratification that its mapper function can return something not in categories? I'm pleasantly surprised if this is actually what happens.

File: tests/framework/results/test_manager.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+from .mocks import CATEGORIES, NAME, SOURCES, sorting_hat_serial, sorting_hat_vector
+
+
+# Mock for get_value call for Pipelines, returns a str instead of a Pipeline
+def mock_get_value(self, name: str):
+    return name
+
+
+def verify_stratification_added(mgr, name, categories, sources, mapper, is_vectorized):
```
Comment (rmudambi) [inline tests/framework/results/test_manager.py:15]: I misread this and thought it was a test. Maybe add a comment to clarify it is not. 

Also, consider whether we want to define the `__eq__` function on `Stratification` to simplify this check. 

In `test_context.py` line 41 you make the assertion `assert expected_object in ctx._stratifications`, and I thought that `in` keyword is implicitly doing an equality check, so check if the default equality function is sufficient for this purpose and that the above test is correct.

File: tests/framework/results/test_manager.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        (  # expected Stratification for non-vectorized
+            NAME,
+            SOURCES,
+            CATEGORIES,
+            sorting_hat_serial,
+            False,
+        ),
+    ],
+)
+def test_register_stratification_with_pipelines(
```
Comment (rmudambi) [inline tests/framework/results/test_manager.py:81]: Add a test where you register a stratification with both column and pipeline sources.

File: tests/framework/results/test_context.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        (  # is_vectorized=False with vectorized mapper
+            NAME,
+            SOURCES,
+            CATEGORIES,
+            sorting_hat_vector,
+            False,
+            Exception,
+        ),
+    ],
+)
```
Comment (rmudambi) [inline tests/framework/results/test_context.py:80]: It occurs to me that it should probably throw an error if someone tries to create a Stratification with duplicate categories - i.e. `categories=['foo', 'bar', 'foo']`. 

Let's create another ticket to handle that edge case and the relevant test case.

**Review by rmudambi — APPROVED** [review]

File: tests/framework/results/test_context.py (hunk 14 above)
Comment (mattkappel) [reply tests/framework/results/test_context.py:80]: Created MIC-3475

File: tests/framework/results/test_manager.py (hunk 12 above)
Comment (mattkappel) [reply tests/framework/results/test_manager.py:15]: I recall why I did it the two different ways, but in retrospect, I don't think my reasoning was right. I'm going to refactor the more deliberate `verify_stratification_added` to be used by both `test_manager.py` and `test_context.py`

File: tests/framework/results/test_context.py (hunk 10 above)
Comment (mattkappel) [reply tests/framework/results/test_context.py]: I removed these tests, as they were nonsensical and replaced it with the only real exception testing I could do for this bit.

File: tests/framework/results/test_context.py (hunk 11 above)
Comment (mattkappel) [reply tests/framework/results/test_context.py]: Yeah, they were nonsense. I don't know why they passed though. Nevertheless, they've been removed and replaced.
